agents/heuristics_resource/fence_heuristics_resource.py

```diff
@@ -0,0 +1,160 @@
+#!/usr/libexec/platform-python -tt
+
+import io
+import re
+import subprocess
+import shlex
+import sys, stat
+import logging
+import atexit
+import time
+sys.path.append("/usr/share/fence")
+from fencing import fail_usage, run_command, fence_action, all_opt
+from fencing import atexit_handler, check_input, process_input, show_docs
+from fencing import run_delay
+
+def heuristics_resource(con, options):
+	# Search the node where the resource is running and determine
+	# the ACT node or not. For SBY node, a delay is generated.
+	# Note that this method always returns FALSE.
+
+	if not "--nodename" in options or options["--nodename"] == "":
+		logging.error("nodename parameter required")
+		return False
+
+	if not "--resource" in options or options["--resource"] == "":
+		logging.error("resource parameter required")
+		return False
+
+	target = options["--nodename"]
+	resource = options["--resource"]
+	promotable = options["--promotable"] in ["", "1"]
+	standby_wait = int(options["--standby-wait"])
+	crm_resource_path = options["--crm-resource-path"]
+	crm_node_path = options["--crm-node-path"]
+
+	(rc, out, err) = run_command(options, "%s --name" % crm_node_path)
+	if rc != 0 or out == None:
+		logging.error("Can not get my nodename. rc=%s, stderr=%s" % (rc, err))
+		return False
+
+	mynodename = out.strip()
+
+	if mynodename == target:
+		logging.info("Skip standby wait due to self-fencing.")
+		return False
+
+	(rc, out, err) = run_command(options, "%s -r %s -W" % (crm_resource_path, resource))
+	if rc != 0 or out == None:
+		logging.error("Command failed. rc=%s, stderr=%s" % (rc, err))
+		return False
+
+	search_str = re.compile(r"\s%s%s$" % (mynodename, '\sMaster' if promotable else ''))
```

Comment:
Have you tested this? I think there's a space at the end of the line if "Master" isn't printed.

I'd be a little concerned about introducing a dependency on the output format. We regularly have problems crop up when other software (e.g. pcs) looks for specific strings in the output, since we don't always remember that something requires it when we need to update the pacemaker output for some reason.

We actually are in the middle of a project to have all the pacemaker tools support XML output, so external code can use a format that we will try to keep backward-compatible while still allowing us to change text output as needed. Unfortunately crm_resource has not yet been adapted, and you'd want to support older pacemaker versions anyway.

Another concern I have is that crm_resource -W doesn't distinguish between running cleanly and failed. The output would be the same in either case. If the resource is failed and the node needs to be fenced, this will unnecessarily delay recovery.

One possibility would be to use crm_mon --as-xml instead, which would avoid both the magic string problem and the success vs failure problem. crm_mon's XML output has been supported "forever". It will generate a bunch of output including something like:

    <resource id="R" ... failed="true" failure_ignored="false" ... >
        <node name="N" ... />
    </resource>

Another corner case is that a resource can be (wrongly) active on more than one node. In a 2-node cluster, it would be simple to skip the delay here if the resource is active on the local node, even if it is also active on the other node. In a larger cluster the best course of action is not as obvious, but a safe choice would be to skip the delay if the resource is active on multiple nodes.

Side note: the upcoming release of pacemaker will have a slight change in how crm_mon does XML output. --as-xml will still behave the same, but it will be deprecated (however it will remain supported for years to come). A new --output-as=xml option will generate nearly identical XML output, but with a different outermost tag. That is part of the project to make all tools support XML -- that option and outermost tag will be identical for all tools.

Reply:
> I think there's a space at the end of the line if "Master" isn't printed.

At this point, it was an implementation that removed the space by `strip()` .

> One possibility would be to use crm_mon --as-xml instead, which would avoid both the magic string problem and the success vs failure problem. crm_mon's XML output has been supported "forever". It will generate a bunch of output including something like:

I added new commit to use `crm_mon --as-xml`.
Is it necessary and sufficient to detect resource failure by checking the following two attributes?

```
failed="true" failure_ignored="false"
```

If even one of the resources has failed, including resources inside the resource group or promotable resource, the delay is skipped.

> Another corner case is that a resource can be (wrongly) active on more than one node. In a 2-node cluster, it would be simple to skip the delay here if the resource is active on the local node, even if it is also active on the other node. In a larger cluster the best course of action is not as obvious, but a safe choice would be to skip the delay if the resource is active on multiple nodes.

Regardless of the number of nodes in the cluster, the delay is skipped when the resource is started on multiple nodes.

In the latest commit, the delay occurs only when the specified resource, resource group, promotable master resource is started on another single node (not remote) and has not failed. The delay is also skipped for self-fencing.